### DISCUSSION TITLE
[QUESTION] kvs_gstreamer_sample hanging after "pad link successful"

### DISCUSSION THREAD

**Discussion by epietrowicz** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#1136] [Q&A] [answered]
I have an RTSP stream from a IR thermal camera that I am trying to push to KVS. I am having trouble getting the kvs_gstreamer_sample working with this camera specifically.

The RTSP stream connects fine with VLC player:

![VLC Stream](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/assets/8181497/6c0c154c-2999-4925-9386-330e60257c9c)

![VLC Codec Information of Thermal Camera](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/assets/8181497/e2814127-067f-4e84-aa95-16d95e449f8e)

I believe my Kinesis stream is configured correctly as I can get it to stream from a different visible light camera. 

I start the sample app with the following command (from the build folder):
```
./kvs_gstreamer_sample camera_1 "rtsp://192.168.0.184/net0"
```

Running the above command on the thermal camera hangs after the `Pad link successful` output.

The logs are exactly the same with my visible light camera, however, after `Pad link successful` I see streaming output from the `PutMedia` requests. 

Then, I am able to view the visible light camera's stream in my AWS console.

[Log output](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/files/14185743/output.txt)
[Log output with GST_DEBUG=4](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/files/14185742/output_debug.txt)

Some things to note:
1. My visible light camera is 60fps (working with KVS) while my thermal camera is 30fps (NOT working)
2. Visible light camera is 1920px x 1080px (working) and thermal camera is 640px x 480px (not working)
3. Both streams are H264 encoded
4. My producer application is running on a Raspberry Pi 4

Could the thermal camera's lower frame rate be causing issues? The resolution? Grayscale versus RGB?

I have tried using gstreamer directly using the following command found [here](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/master/docs/linux.md#running-the-gst-launch-10-command-to-start-streaming-from-a-rtsp-camera-source):
```
gst-launch-1.0 -v rtspsrc location="rtsp://192.168.0.184/net0" short-header=TRUE ! rtph264depay ! h264parse ! kvssink stream-name="camera_1" iot-certificate="iot-certificate,endpoint=<my_endpoint>,cert-path=certs/certificate.pem.crt,key-path=certs/private.pem.key,ca-path=certs/AmazonRootCA1.pem,role-aliases=<My_role_alias>"
```

The above command results in a continuous stream of the following on my thermal camera:
```
[DEBUG] [06-02-2024 20:36:52:393.044 GMT] Dropping frame with flag: 9728 for camera_1
```

[Full output from gst-launch command](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/files/14185825/raw_output.txt)

Any ideas on what else I can try?

**Comment by niyatim23** [comment]
Can you try using the sample with filesink instead of kvssink and see if the issue persists? Also, have you tried modifying the sample application's pipeline with the one that works for you?

**Reply by epietrowicz** [reply]
Hi @niyatim23,
Thank you for your response. Filesink works fine for both cameras with the following pipeline:

```
gst-launch-1.0 -v rtspsrc location=rtsp://192.168.0.184/net0  latency=300 ! rtph264depay ! h264parse ! mpegtsmux ! filesink location=output_video.mpeg
```

How should I update the sample pipeline in this case? It works well with filesink. Is there some additional flag I can pass into the kvssink plugin?

With kvssink, the pipeline is repeatedly logging the following error: 
```
[DEBUG] [07-02-2024 15:53:35:490.304 GMT] Dropping frame with flag: 9728 for camera_1
```

The kvssink plugin pipeline (same as before):
```
gst-launch-1.0 -v rtspsrc location=rtsp://192.168.0.184/net0 latency=300 ! rtph264depay ! h264parse ! kvssink stream-name="camera_1" iot-certificate="iot-certificate,endpoint=***,cert-path=certs/certificate.pem.crt,key-path=certs/private.pem.key,ca-path=certs/AmazonRootCA1.pem,role-aliases=***"
```

**Comment by epietrowicz** [comment, marked as answer]
Commenting out `!GST_BUFFER_DTS_IS_VALID(buf)` and rebuilding the kvssink plugin as described in [this solution](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/discussions/1008#discussioncomment-5850733) solved the issue for us.
